### DISCUSSION TITLE
DietPi-Software | Caddy server

### DISCUSSION THREAD

**Discussion by jcw** [MichaIng/DietPi#5370] [Software Request]
#### Formal software information
- Software title | Caddy server
- Short description | self-contained web- and proxy server
- Official URL | https://caddyserver.com
- GitHub URL | https://github.com/caddyserver/caddy
- (Official) install documentation | https://caddyserver.com/docs/install

#### Are there similar/alternative software titles available with DietPi-Software?
- Nginx, Apache, etc

#### What makes your requested software better than the above solutions, if available?
- does HTTPS (w/ Let's Encrypt) by default, for multiplel domains
- very simple "Caddyfile" config when used for simple configurations
- a single executable, and a single config file (unless you prefer to split it)
- built in Go, which has its own strong and independent network stack (with Google behind it)

#### How can DietPi make the installation easier or compatible, than following the install instructions or do APT installation, if available?
- a DIY install of caddy is easy, but it interferes with other web servers on ports 80 & 443
- the key benefit (auto-HTTPS) of caddy mainly adds value if it were made the "central" web server
- this is of course not going to satisfy everyone, so "switching to" caddy probably needs to be optional
- but this is already supported in DietPi, with the choice of a "Webserver Preference" in dietpi-software

#### Can you provide the installation steps that you would suggest DietPi-Software to do?
1. Install Caddy (single-exe, just fetch the proper one from github release page)
2. prepare minimal CaddyFile (the usual path is /etc/caddy/, I think)
3. add new "caddy.service" to systemd, see https://caddyserver.com/docs/install#linux-service

#### Are you willing to help maintaining the software installation, e.g. in case of needed setup changes due to updates etc.? This is not needed, but could speed up our decision to implement it, as man power is always a topic :wink:.
- I can help, in that I'm fairly familiar with Go, and have been using Caddy for some time (v2 only since mid 2020)
- I've set it up for static sites, as PHP proxy (e.g. Nextcloud), and as reverse proxy (e.g. Gitea)
- and FWIW: no "deep" insight into Caddy, moderate networking knowledge, well-versed in Unix stuff

#### Vote for this software on FeatHub: https://feathub.com/MichaIng/DietPi/+100

(related issue: #1622)

**Comment by gt2416** [comment]
+1 for this. Caddy automatically manages certificates and setting it up as a reverse proxy is so much easier than using Apache / nginx.

**Reply by faisalbasra** [reply]
+1

**Comment by jalsco** [comment]
Hmmm - I been using Caddy Server on DeitPi Buster since August 2021. It works flawlessly (and it is awesome 😁✨) .  I agree with the other posters that if it were in the autoinstall packages that would be better.  As now it is a manual install - which is very straightforward (just follow the steps on the Caddy website for Debian ) .

I was checking in here in case reported problems running caddy on the new Deipi Pi (Debian 11) . I did not see any issues reported - so I will continue with the DietPi upgrade.  Assuming Caddy runs then I will not post again. However if it fails - I will come back and edit this post  and also downgrade by DietPi back to 10-Buster... But I expect it will work !

**Comment by Xerrion** [comment]
+1 on this, have been using Caddy for over a year on my dietpi, would be really neat to have this available built-in as an alternative.

**Comment by eikaramba** [comment]
+1 too - honestly i find caddy so much easier than nginx/lighttpd

for the others: is it a straight install on the dietpi debian 11? or anything to look out for?

**Reply by Joulinar** [reply]
see my comment below. Quite easy and straightforward.

**Comment by jcw** [comment]
Yep, should just work - there's no install, it's a single executable.

Meanwhile, I'm about to fall back to Nginx + CertBot for a new server I'm preparing here, given the lack of support in DietPi.
Unfortunate, but not the end of the world. Still, as others point out, Caddy is a breath of fresh air. Wink, wink, @MichaIng :)

**Comment by gt2416** [comment]
Nothing special to do. You can build it on diet pi easily. You have to install go first to build it. Works as you would expect

**Reply by Joulinar** [reply]
why are you trying to build it from source? Caddy is offering ready to use Debian packages *.deb files directly on their GitHub. Depending on your OS type, simply download `linux_arm6/7/64.deb` and use `dpkg -i <file.name.deb>` to install

It can't be more simple 😉 

```
root@DietPi3:~# cd /tmp/
root@DietPi3:/tmp# wget https://github.com/caddyserver/caddy/releases/download/v2.4.6/caddy_2.4.6_linux_armv6.deb
--2022-03-19 21:19:48--  https://github.com/caddyserver/caddy/releases/download/v2.4.6/caddy_2.4.6_linux_armv6.deb
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Saving to: ‘caddy_2.4.6_linux_armv6.deb’

caddy_2.4.6_linux_armv6.deb                  100%[==============================================================================================>]  10.62M  5.43MB/s    in 2.0s

2022-03-19 21:19:51 (5.43 MB/s) - ‘caddy_2.4.6_linux_armv6.deb’ saved [11135780/11135780]
```

```
root@DietPi3:/tmp# dpkg -i caddy_2.4.6_linux_armv6.deb
Selecting previously unselected package caddy.
(Reading database ... 22513 files and directories currently installed.)
Preparing to unpack caddy_2.4.6_linux_armv6.deb ...
Unpacking caddy (2.4.6) ...
Setting up caddy (2.4.6) ...
Created symlink /etc/systemd/system/multi-user.target.wants/caddy.service → /lib/systemd/system/caddy.service.
```

```
root@DietPi3:/tmp# systemctl status caddy
● caddy.service - Caddy
     Loaded: loaded (/lib/systemd/system/caddy.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2022-03-19 21:20:02 CET; 15s ago
       Docs: https://caddyserver.com/docs/
   Main PID: 755 (caddy)
      Tasks: 9 (limit: 2178)
        CPU: 147ms
     CGroup: /system.slice/caddy.service
             └─755 /usr/bin/caddy run --environ --config /etc/caddy/Caddyfile

Mar 19 21:20:02 DietPi3 caddy[755]: JOURNAL_STREAM=8:11880
Mar 19 21:20:02 DietPi3 caddy[755]: {"level":"info","ts":1647721202.1006916,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":""}
Mar 19 21:20:02 DietPi3 caddy[755]: {"level":"info","ts":1647721202.1070118,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["[::1]:2019","127.0.0.1:2019","localhost:2019"]}
Mar 19 21:20:02 DietPi3 caddy[755]: {"level":"info","ts":1647721202.1075172,"logger":"http","msg":"server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server","server_name":"srv0","http_port":80}
Mar 19 21:20:02 DietPi3 caddy[755]: {"level":"info","ts":1647721202.1076684,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x2b837c0"}
Mar 19 21:20:02 DietPi3 caddy[755]: {"level":"info","ts":1647721202.1085865,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/var/lib/caddy/.local/share/caddy"}
Mar 19 21:20:02 DietPi3 caddy[755]: {"level":"info","ts":1647721202.1091156,"logger":"tls","msg":"finished cleaning storage units"}
Mar 19 21:20:02 DietPi3 caddy[755]: {"level":"info","ts":1647721202.109399,"msg":"autosaved config (load with --resume flag)","file":"/var/lib/caddy/.config/caddy/autosave.json"}
Mar 19 21:20:02 DietPi3 caddy[755]: {"level":"info","ts":1647721202.1096168,"msg":"serving initial configuration"}
Mar 19 21:20:02 DietPi3 systemd[1]: Started Caddy.
```

```
root@DietPi3:/tmp# ss -tulpn | grep LISTEN
tcp   LISTEN 0      4096       127.0.0.1:2019      0.0.0.0:*    users:(("caddy",pid=755,fd=3))
tcp   LISTEN 0      1000         0.0.0.0:22        0.0.0.0:*    users:(("dropbear",pid=321,fd=4))
tcp   LISTEN 0      4096               *:80              *:*    users:(("caddy",pid=755,fd=7))
tcp   LISTEN 0      1000            [::]:22           [::]:*    users:(("dropbear",pid=321,fd=5))
root@DietPi3:/tmp#
```

**Reply by gt2416** [reply]
> why are you trying to build it from source? 

For plugins.

**Reply by anubis-genix** [reply]
What are the chances to see caddy or another reverse proxy being implemented in DietPi?

**Reply by Joulinar** [reply]
Well we don't have time to implement all software title that are requested by users. Therefore it might take a while. If someone find time to start implementing, volunteers are welcome 😀

**Reply by jalsco** [reply]
it already works perfectly without an explicit "dietpi installation menu item". 
I love caddyserver - but honestly nothing is needed to use it on Dietpi. 
People can use it now - I been doing so for over a year.

**Comment by ghost** [comment]
+1

**Comment by Ricardo-Simoes** [comment]
+1 Caddy Server is light and has Auto SSL certificates

**Comment by oldboys92** [comment]
+1 ACME is supported without a certbot, it is feature inside the Caddy server. Love it and is so easy to setup a reverse proxy with Caddy including ACME even for a private CA (fe. Step CA).

Never the less it is very easy to install it manually via apt package manager.
